### ISSUE TITLE
Linux auto updater (update when Firefox starts)

### ISSUE THREAD

**Issue by ShellCode33** [arkenfox/user.js#1696]
Dear fellow Linux users,

I think we can all agree that running the updater script manually is cumbersome.
I tend to forget things, and updating the user.js is for sure on my top 5 of things to forget.
That's why I try to automate as much stuff as possible and arkenfox user.js is no exception.
Allow me to share with you a small hack I did to automate everything.

First of all, and this is very important, my PATH variable looks like this:

```
/usr/local/bin:/usr/bin:/home/shellcode/.local/bin
```
Notice how `/usr/local/bin` is **before** `/usr/bin`.

Then I created a `firefox` script in `/usr/local/bin` with the following content:

<details>
  <summary><b>[CLICK TO SHOW SCRIPT]</b></summary>

```python
#!/usr/bin/python3

import os
import re
import sys
import os.path
import requests
from typing import List, Tuple
from configparser import ConfigParser

FIREFOX_DIR = os.environ["HOME"] + "/.mozilla/firefox"
USER_JS_OVERRIDES_PATH = f"{FIREFOX_DIR}/user-overrides.js"

USER_JS_URL = "https://raw.githubusercontent.com/arkenfox/user.js/master/user.js"
USER_PREF_REGEX = r'user_pref\("(?P<pref_key>.+)"'

def get_profiles() -> List[Tuple[str, str]]:

    profiles = []

    config = ConfigParser()
    config.read(FIREFOX_DIR + "/profiles.ini")
    
    for section in config.sections():
        if section.startswith("Profile"):
            name = config[section]["Name"]
            path = config[section]["Path"]
            profiles.append((name, path))

    return profiles

def install(profile_path: str, new_user_js_content: str) -> None:
    
    user_js = f"{profile_path}/user.js"

    with open(user_js, "w") as stream:
        stream.write(new_user_js_content)

        with open(USER_JS_OVERRIDES_PATH, "r") as stream_override:
            stream.write("\n\n// ------ OVERRIDES START HERE ------\n\n")
            stream.write(stream_override.read())

def clean(profile_path: str) -> None:
    """
    Remove all entries from prefs.js that are in user.js regardless of whether they are active or not.
    They will be set back from user.js next time Firefox starts.

    Firefox must be closed for this to work because prefs.js is overwritten on exit.
    """
    user_js = f"{profile_path}/user.js"
    prefs_js = f"{profile_path}/prefs.js"

    # prefs.js doesnt exist, it might be an unused profile
    if not os.path.exists(prefs_js):
        return

    prefs_to_remove = []

    with open(user_js, "r") as stream:
        for line in stream:
            match = re.search(USER_PREF_REGEX, line)

            if match:
                prefs_to_remove.append(match.group("pref_key"))

    with open(prefs_js, "r") as stream:
        prefs_js_content = stream.read()

    with open(prefs_js, "w") as stream:
        for line in prefs_js_content.split("\n"):
            match = re.match(USER_PREF_REGEX, line)

            if match and match.group("pref_key") not in prefs_to_remove:
                stream.write(f"{line}\n")

def main() -> None:
    try:
        print(f"Downloading {USER_JS_URL}")
        new_user_js_content = requests.get(USER_JS_URL, timeout=0.5).text

        for profile_name, profile_path in get_profiles():
            profile_full_path = f"{FIREFOX_DIR}/{profile_path}"
            print(f"Processing profile: {profile_name}")
            install(profile_full_path, new_user_js_content)
            clean(profile_full_path)

    except requests.exceptions.Timeout:
        "Timeout reached, internet is probably not available, but it's ok, let's start Firefox anyway"

    # Run the real firefox
    os.execv("/usr/bin/firefox", ["firefox"])

if __name__ == "__main__":
    main()
```

</details>

(Don't forget to `sudo chmod +x /usr/local/bin/firefox`)

The idea is that when you run Firefox, the wrapper script `/usr/local/bin/firefox` will be executed instead of the real one (`/usr/bin/firefox`).
That way we can update `user.js` and apply our `user-overrides.js` automatically during each startup. We can even do what `prefsCleaner.sh` does because Firefox is not started yet ! Note that my script does everything by itself, no `updater.sh` or `prefsCleaner.sh` is required.

By default this script expects the overrides to be located at `~/.mozilla/firefox/user-overrides.js`.

You don't have to do this but personally, I have a symlink to my dotfiles:

```
ln -s ~/.dotfiles/.mozilla/firefox/user-overrides.js ~/.mozilla/firefox/user-overrides.js
```

**Warning: my script applies arkenfox user.js and the overrides to all the profiles.** If this is not what you want, just change the script to your own liking.

Here are some questions I asked myself and you could ask yourself as well:

------------

**What happens if I forget to upgrade Firefox itself using my package manager ?**

Worst case scenario, your user.js will be more up to date than Firefox. But I tested this and Firefox doesn't seem to complain when user prefs it doesn't know about are set.

Try it yourself, add something like `user_pref("ThisIsSomethingRandomThatDoesntExist", false);` to your user.js and restart Firefox.

------------

**What happens if an instance of Firefox is already running and I start a new one ?**

Both the `user.js` and `prefs.js` will be updated but the changes to `prefs.js` will be reverted because Firefox writes its runtime configuration to `prefs.js` on exit. But this is completely fine ! It will just get updated for real the next time you completely close and reopen Firefox.

------------

Sorry for hijacking your issue tracker, I just think it will get more visibility here. Closing now so that you don't have to :-)

**Comment by Thorin-Oakenpants** [comment]
> Firefox doesn't seem to complain when user prefs it doesn't know about

pref names with typos (or wrong case), or indeed even MADE UP names like the parrot have zero effect

**Comment by carpet-stain** [comment]
@ShellCode33 @Thorin-Oakenpants 

Is there a way to adapt this script to macOS? As far as I know, macOS does not use a predictable profile location. I wonder if there is a way to figure this directory in an automated way without having to open `about:support` and copying the directory.

**Comment by Thorin-Oakenpants** [comment]
not my field sorry .. but the fish might know since he comes from an elite rich and powerful school who could afford all the shiny overpriced apple things @fxbrit

**Comment by ShellCode33** [comment]
I'm not sure what you mean by

> As far as I know, macOS does not use a predictable profile location

I'm not a MacOS user but according to [this](https://kb.mozillazine.org/Profile_folder_-_Firefox) (granted it's not official), the profile folder is in one of the two following locations:

- `~/Library/Application Support/Firefox/Profiles/<profile folder>`
- `~/Library/Mozilla/Firefox/Profiles/<profile folder>`

What does your `about:support` say ?

**Comment by fxbrit** [comment]
I agree with @ShellCode33, `~/Library/Application Support/Firefox/Profiles/` is the standard profile directory unless you are doing some oddity like running the app inside the disk image or installing it outside of the application folder. In this last case the profile directory will be in `path/to/the/app/Firefox-Data/Browser`.

**Comment by ShellCode33** [comment]
I don't have time to do this right now but it would be nice to port the idea of my script to other platforms (both Windows and MacOS). IMO auto-updating is a far better approach than running scripts manually (Im pretty sure 99% of users forget to update their user.js when Firefox is being updated, especially on Windows where it is updated automatically in the background). On Windows, users could be instructed in the wiki to replace their desktop shortcut to point to a script, and on MacOS I guess the `$PATH` trick could work as well ? I don't know how apps are being launch on MacOS (graphically speaking). It is very platform dependent and there are probably some edge cases to avoid. For example on Windows, does the desktop shortcut get updated when Firefox is updated ? Will it override the shortcut pointing to the script ? I have no idea.

**Comment by 9ao9ai9ar** [comment]
@ShellCode33 I agree with you that shell scripting sucks (see my work in progress pull request linked above for the hurdles I have to overcome to make the scripts portable). Your approach has merits, and choosing to implement it in Python 3 is sound because it is installed by default on every platform I have tested upon installing Firefox (except NixOS), and, due to being a [build time dependency](https://firefox-source-docs.mozilla.org/build/buildsystem/python.html), it needs to be available anyway if one chooses to build Firefox from source. There are a few issues, however, that you need to take into consideration:

1. Python version compatibility. The default Python installations on the platforms I have tested range from Python 3.6 (openSUSE Leap 15.6) / Python 3.7 (Oracle Solaris 11.4) to Python 3.12, with `python3.X` and the not always available symbolic links, `python3` and `python`, present on `PATH`. I suspect the reason why the symbolic links are not provided on some platforms is because Python is very bad at backward compatibility (see [this Stack Overflow post](https://stackoverflow.com/q/70467517) and [what Red Hat had to fix by upgrading to Python 3.9 from Python 3.8](https://bugzilla.redhat.com/show_bug.cgi?id=1785415) for more details) and they want to make sure all Python scripts are written and tested for `python3.X`.
2. Firefox profile path. On Ubuntu 24.04 LTS, the default provided Firefox is a Snap app and its profile location is `~/snap/firefox/common/.mozilla/firefox`, and on Haiku (Iceweasel) it is `~/config/settings/Mozilla/Firefox`. I'm sure there could be more locations for the profile root (e.g. if it's installed as a Flatpak app), so the script should be configurable to allow for user-provided path for the Firefox profile, just like the old updater script (whether through the `-p` option or auto-detected based on the path of the script).
